### PR TITLE
Feature #2780 grib tables

### LINKED ISSUE #2780: Update GRIB tables in MET based on wgrib2 versions 3.1.4 and 3.4.0
## Describe the Enhancement ##
Note that this work was performed using the GRIB2 table input from:
- wgrib2 version 3.1.4 gribtables (a separate development branch prior to the transition to GitHub) from:
  `hera:/home/Wesley.Ebisuzaki/grib2/wgrib2/gribtables`
- wgrib2 version 3.4.0 [gribtables](https://github.com/NOAA-EMC/wgrib2/tree/v3.4.0/wgrib2/gribtables)
  - See NOAA-EMC/wgrib2#285 for some of the issues/discrepancies I found.

This issue was discussed during the METplus NOAA User Telecon on Jan 8, 2024 (see [meeting notes](https://docs.google.com/document/d/1L6fEUuLsb4JpZflZZe52Ixmm9To4AqiZSx0kadF28cA/edit) - limited access). Gwen indicated that [wgrib2](https://www.cpc.ncep.noaa.gov/products/wesley/wgrib2/) version 3.1.4 is under development and will include updates to some of the GRIB2 tables. Note that wgrib2 has migrated to GitHub at [NOAA-EMC/wgrib2](https://github.com/NOAA-EMC/wgrib2).

> [NCEP GRIB2 tables](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-2.shtml) have been updated with new WMO-approved variables (27 freezing/melting potential, 28 melt onset date, and 29 freeze onset date in [Table 4.2-10-2](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-2-10-2.shtml) and Level 172 in [Table 4.5](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-5.shtml)).

> A new release v3.1.4 will be installed on Acorn soon for testing. If successful, it will be delivered to WCOSS2.
> Use the v3.4.0 release from 8/15/24 instead.

This issue is to:
- Download the v3.1.4 release when it becomes available.
- Update the contents of the GRIB2 table files in `data/table_files/grib2_*.txt` as needed.

Recommend requesting that Gwen (Li-Chuan Chen) serve as the tester for these table updates. 

### Time Estimate ###
1 day.

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

This PR makes changes to GRIB1 and GRIB2 table files in MET. These changes include:
- Incorporating new tables and modified entries from wgrib2 version 3.1.4 (hera:/home/Wesley.Ebisuzaki/grib2/wgrib2/gribtables) and from wgrib2 version 3.4.0 from NOAA-EMC/wgrib2.
- Note that these version numbers appear sequential but were created from different development branches.
- Eliminating whitespace from the units strings which are written to the `FCST_UNITS` and `OBS_UNITS` STAT header columns, where embedded whitespace is not allowed.
- Modifying units strings from using `**` to `^` to consistently indicate exponentiation (e.g. switch from `m**2` to `m^2`).
- Correct many units strings that lacked `**` or `^` for exponentiation (e.g. switch from `m2` to `m^2`).
- Modifying units strings to replace negative exponents with `/` (e.g. switch from `m s**-1` to `m/s`).
- Modifying units strings to eliminate parentheses in the denominator (e.g. switch from `kg/(m^2*s)` to `kg/m^2/s`).
- I note that some GRIB table abbreviations (in the `grib1_dwd_*.txt` tables) do still include embedded whitespace. MET does check for that when writing the `FCST_VAR` and `OBS_VAR` STAT header columns and reports this error message:
```
ERROR  : check_hdr_str() -> output header column FCST_VAR value ("Temp With Spaces") should contain no embedded whitespace!
```
Users can use `set_attr_name = "TMP"` to provide a non-whitespace version of the name.

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
None yet.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Spend no more than 1 hour scanning the differences, looking for obvious problems in changes to the tables, in particular for updated unit strings.
Please refer to this [NCEP GRIB2 documentation](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-1.shtml) and this [NCEP GRIB1 documentation](https://www.nco.ncep.noaa.gov/pmb/docs/on388/table2.html) only when needed.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**
None needed.

- [x] Do these changes include sufficient testing updates? **[Yes]**
None needed.

- [x] Will this PR result in changes to the MET test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
The vorticity units reported to the MET differ:
```
COMPARING grid_stat/grid_stat_GRIB_lvl_typ_val_180000L_20170610_180000V_pairs.nc
file1: /data/output/met_test_truth/grid_stat/grid_stat_GRIB_lvl_typ_val_180000L_20170610_180000V_pairs.nc
file2: /data/output/met_test_output/grid_stat/grid_stat_GRIB_lvl_typ_val_180000L_20170610_180000V_pairs.nc
ERROR: NetCDF headers differ:
71c71
< 		FCST_PVORT_L310_FULL:units = "Km^2/kg/s" ;
---
> 		FCST_PVORT_L310_FULL:units = "K*m^2/kg/s" ;
```
See the [GRIB2 table documentation](https://www.nco.ncep.noaa.gov/pmb/docs/on388/table2.html#TABLE128) for:
Screenshot 2024-11-01 at 4 22 42 PM

- [x] Will this PR result in changes to existing METplus Use Cases? **[Probably]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.
The units reported to the MET output may differ.

- [x] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:
I expect no impact.

- [x] Please complete this pull request review by **[Tues 11/5/24]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: data/table_files/grib1_cptec_254_64.txt [hunk 1]
```diff
@@ -161,24 +161,24 @@ GRIB1
...
 164 254 46 -1 "gvus" "GRAVITY WAVE DRAG SFC ZONAL STRESS " "Pa"
 165 254 46 -1 "gvvs" "GRAVITY WAVE DRAG SFC MERIDIONAL STRESS" "Pa"
 166 254 46 -1 "var166" "undefined" ""
 167 254 46 -1 "dvsh" "DIVERGENCE OF SPECIFIC HUMIDITY" "1/s"
 168 254 46 -1 "hmfc" "HORIZ. MOISTURE FLUX CONV. " "1/s"
-169 254 46 -1 "vmfl" "VERT. INTEGRATED MOISTURE FLUX CONV." "kg/(m2*s)"
-170 254 46 -1 "vadv" "VERTICAL MOISTURE ADVECTION " "kg/(kg*s)"
-171 254 46 -1 "nhcm" "NEG. HUM. CORR. MOISTURE SOURCE" "kg/(kg*s)"
+169 254 46 -1 "vmfl" "VERT. INTEGRATED MOISTURE FLUX CONV." "kg/m^2/s"
+170 254 46 -1 "vadv" "VERTICAL MOISTURE ADVECTION " "kg/kg/s"
```
Comment (CPKalb) [inline data/table_files/grib1_cptec_254_64.txt:172]: Seconds are usually on the bottom of the fraction for moisture advection.  I think this should say kg/kg*s...

File: data/table_files/grib1_cptec_254_64.txt [hunk 2]
```diff
@@ -161,24 +161,24 @@ GRIB1
...
+163 254 46 -1 "gvdv" "GRAVITY WAVE DRAG DV/DT" "m/s^2"
 164 254 46 -1 "gvus" "GRAVITY WAVE DRAG SFC ZONAL STRESS " "Pa"
 165 254 46 -1 "gvvs" "GRAVITY WAVE DRAG SFC MERIDIONAL STRESS" "Pa"
 166 254 46 -1 "var166" "undefined" ""
 167 254 46 -1 "dvsh" "DIVERGENCE OF SPECIFIC HUMIDITY" "1/s"
 168 254 46 -1 "hmfc" "HORIZ. MOISTURE FLUX CONV. " "1/s"
-169 254 46 -1 "vmfl" "VERT. INTEGRATED MOISTURE FLUX CONV." "kg/(m2*s)"
-170 254 46 -1 "vadv" "VERTICAL MOISTURE ADVECTION " "kg/(kg*s)"
-171 254 46 -1 "nhcm" "NEG. HUM. CORR. MOISTURE SOURCE" "kg/(kg*s)"
+169 254 46 -1 "vmfl" "VERT. INTEGRATED MOISTURE FLUX CONV." "kg/m^2/s"
```
Comment (CPKalb) [inline data/table_files/grib1_cptec_254_64.txt:171]: Was this supposed to change from * to /?  Same for the next 2 lines...

**Review by CPKalb — COMMENTED** [review]
It seems like many of the * have changed to /.  I just wanted to double check that this was intentional...

File: data/table_files/grib1_cptec_254_64.txt (hunk 2 above)
Comment (JohnHalleyGotway) [reply data/table_files/grib1_cptec_254_64.txt:171]: @CPKalb, yes, it was an intentional change... switching from `a/(b*c)` to `a/b/c` which is mathematically equivalent.  I found a pretty wide variety of conventions in units strings across GRIB1 and GRIB2 table sources. And I made these changes to move toward more consistency.

But please advise if you'd prefer that those differences be retained.

Comment (CPKalb) [reply data/table_files/grib1_cptec_254_64.txt:171]: @JohnHalleyGotway, I think the change is fine, I just wanted to be sure.  Like you said, there are a variety of ways to express it.

**Review by CPKalb — APPROVED** [review]
Went through as many of the changes as I could in an hour and didn't see anything amiss.

**Pull request merged**